### ISSUE TITLE
Share accumulated phrase (Telegram, Email,....)

### ISSUE THREAD

**Issue by arasaac-dga** [asterics/AsTeRICS-Grid#294]
Related to #293 (because it would be necessary to create first the image or PDF of the accumulated phrase to share)... Add an option/action to share accumulated phrases by Telegram, Email,....

**Comment by klues** [comment]
First working version with demo config can be found here: https://gridbeta.asterics-foundation.org/?gridset_filename=matrix_messenger_demo.grd.json

Short demo video: https://www.facebook.com/1084147685/videos/1624770688182854/

What's working:
* Connect to an account from [matrix / element](https://app.element.io/#/register) - I've used this platform because it's free, open source, federated and has an open API and so I hope it survives longest. Unfortunately WhatsApp or Signal do not have an open API, so I couldn't use it. Telegram is quite closed, so it also wasn't optimal.
* Create and manage conversations (rooms)
* Create custom chat user interfaces using the new "Matrix messenger" element type
* Send text messages, receive text and picture messages
* auto-read incoming messages

Open:
* UI adaptions (font-size of conversation etc.)
* Sending images - what's the best way of sending the current accumulated phrase - probably a screenshot?!
* voice messages?! - probably I would do this in a separate issue somewhen else

@arasaac-dga @ms-mialingvo if you have time, please test and give me feedback.

**Comment by ms-mialingvo** [comment]
I don't have access to iPads/tablets for the rest of the week so "only" tested betwen laptop and iPhone. It's loking good! :)

> UI adaptions (font-size of conversation etc.)

It looks good to me the way it's currently set up. Although if someone wanted to make changes to that, changing font-size in `appearance` is not impacting the chat?

> Sending images - what's the best way of sending the current accumulated phrase - probably a screenshot?!

The AG user sending images? Don't know what the best way technically is. What I noticed is that if the not-AG-user sends an image and adds a text to the image, that added text isn't shown.

> voice messages?! - probably I would do this in a separate issue somewhen else

Yeah, it's showing just a blank message when the non-AG-user sends something.

**Comment by klues** [comment]
Now it's possible to share the collected pictograms. See this demo:

![Image](https://github.com/user-attachments/assets/77cfe119-9df8-4230-a22a-cc42eb389044)

Other improvements:
* new action to speak last message of conversation
* emoji-only text messages show the emojis in bigger font size
* added loading spinner after sending message and waiting for message to be sent

> Although if someone wanted to make changes to that, changing font-size in appearance is not impacting the chat?

Currently not. I could do it, but maybe it makes more sense to have options like font-size in the edit dialog of the messenger element?!

> The AG user sending images? Don't know what the best way technically is. 

I meant sending the collected phrase + pictograms - see the demo above.

> hat I noticed is that if the not-AG-user sends an image and adds a text to the image, that added text isn't shown.

It now shows the name of the image as text below the image (= the collected phrase if the image was sent with AG)

Regarding voice messages and other advanced features I think it would make most sense to add new issues for it (or let users create issues), if it's really needed. I think the current version is a good basis and we'll see how many users actually use this new feature.

So from my point of view it's nearly finished and I'll release it soon. Please test and tell me if you find any bugs or important improvements before releasing.

Updated demo configuration with new `send collected pictograms` page: https://gridbeta.asterics-foundation.org/?gridset_filename=matrix_messenger_demo.grd.json
(you have to open in inkognito tab or delete the existing demo configuration before in order to update it)

**Comment by ms-mialingvo** [comment]
> https://gridbeta.asterics-foundation.org/?gridset_filename=matrix_messenger_demo.grd.json (you have to open in inkognito tab or delete the existing demo configuration before in order to update it)

It's supposed to look like in the pic, right? It's showing me the keyboard in incognito mode too, but maybe emptying cache or something helps. Anyway, I lost my mobile phone and probably won't be able to test properly until Friday.

**Comment by ms-mialingvo** [comment]
> Currently not. I could do it, but maybe it makes more sense to have options like font-size in the edit dialog of the messenger element?!

I mean, as long as someone doesn't specifically ask for it, any separate option instead of using already existing options adds complexity for the user.

**Comment by arasaac-dga** [comment]
Sorry for the delay in the reply but we were on easter holidays until today. 

Before discussing the new implementation of chat through Matrix, we'd like to mention that, now that it's possible to send a screenshot with the complete phrase in image mode via chat, we could also address issue #293, in which we requested a new option for a saved phrase that would allow users to save it as an image or PDF, or copy it to the clipboard. This is a highly requested option by our users, and we'd appreciate your implementation in conjunction with the development of the chat.

Regarding issue #294, our original idea wasn't so much to include the chat itself within the communicator (which we also consider positive for some users), but rather to have a Share option, like other communicators and many websites have, that would allow the phrase to be sent in image mode to one of the services available on the device (WhatsApp, Telegram, Email, etc.). It would be similar to what you see in this screenshot and what appears, especially, on Android and iOS mobile devices.

![Image](https://github.com/user-attachments/assets/51f9a7f6-c946-4e61-a55a-b116a69e86e5)

In this way, a share action could provide an intermediate solution for those users who use WhatsApp or Telegram and, due to the API difficulties you mentioned, are unable to connect and include chats within Asterics. It would also connect with other social media services such as Instagram, Facebook, and email.

Regarding the Chat integration through Matrix, we've tested it and it works correctly, although it took us several attempts and tests to get it working, mainly due to the fact that encrypted conversations (rooms) didn't work, and we didn't know why. When we created a new room directly from Asterics and saw that it was created unencrypted, we determined that this was the problem.

Today we tried sending phrases, and it works correctly, although we see a small problem in that the image it creates has a lot of gray space when the phrase is short. We don't know if it would be possible to adjust the size of the capture to that of the phrase in some way so that the image only reflects the space used by the phrase.

![Image](https://github.com/user-attachments/assets/263e8630-16ef-40cf-9b07-f0a3918d0731)

Other options like emoticons work perfectly.

**Comment by ms-mialingvo** [comment]
I tested now on iOS, Windows and Android, I had no issues with room creation. 
Microsoft Edge (on Windows and Android tablet) doesn't let me log in, like there's the rotating circle next to the `login` and it stays. It's working in the Firefox/Chrome/Samsung browsers.
Seconding the "issue" with the grey space and the rest arasaac-dga has written.

**Comment by klues** [comment]
Thanks for the feedback!

Changed this:
* background of shared image is white and width only as wide as the content - also better if content too long - still shows all elements in the screenshot
* added two new actions for "Collect element action":
   * `Share as image`: shares the current sentence as image, so can be opened in another app/programm or sent with WhatsApp etc.
   * `Copy image to clipboard`: copies the current sentence as image to clipboard
* `Ctrl + C` (or `Cmd + C`) in normal view mode of a grid - also copies the current sentence as image to clipboard

Please test the new functionalities, then I'll release everything soon.
@arasaac-dga please translate everything in crowdin - thanks!

**Comment by klues** [comment]
Font size of matrix chat now relates to `Font size (text + image)` from normal font size settings. If it's needed to be configurable separately, i would do it in the settings for the matrix messenger element some day. But for now I would leave it as-is.

> Microsoft Edge (on Windows and Android tablet) doesn't let me log in, like there's the rotating circle next to the login and it stays. It's working in the Firefox/Chrome/Samsung browsers.

I tried it on Edge / Windows - for me it worked 🤷

**Comment by arasaac-dga** [comment]
> [@arasaac-dga](https://github.com/arasaac-dga) please translate everything in crowdin - thanks!

Made it!!!

> Changed this:
> 
> * background of shared image is white and width only as wide as the content - also better if content too long - still shows all elements in the screenshot
> * added two new actions for "Collect element action":
>   
>   * `Share as image`: shares the current sentence as image, so can be opened in another app/programm or sent with WhatsApp etc.
>   * `Copy image to clipboard`: copies the current sentence as image to clipboard
> * `Ctrl + C` (or `Cmd + C`) in normal view mode of a grid - also copies the current sentence as image to clipboard

Tested and working as expected. Great!!!

We have found a little bug with our "automatic grammar correction" in Spanish in all the actions that need to apply it. Actually the system doesn't fix the phrases before to send them to the chat or copy to the clipboard or share it. In send to chat action, it is essential that work because it's a conversation with a third person and natural language is important for better communication. 
On the other hand in the new "Share as image" and "Copy image to clipboard" we have tried to add two "Collect element actions" in, for example, the standard PLAY button adding "Copy image to clipboard" (as a second action) to the actual "Speak collect element content" but due that the "automatic grammar correction" send the phrase to our server and spend some ms/s in come back fixed the "Copy image to clipboard" copy the phrase without fix. So, in this case of people that want to combine both actions will not work. 

If you add a separate button to "Share as image" and "Copy image to clipboard" on board and previously press on PLAY and after hearing the phrase fixed you press Share o Copy, of course, there is no problem and share/copy them fixed. 
So, we have the doubt which would be better for "Share as image" and "Copy image to clipboard" (if it would be necessary that this actions call to automatic grammar correction before to copy or share the phrase) because we are not sure if people will use always as a separate button from PLAY (but they will have to press previously it to correct the phrase) or want to combine both actions in only want button. 

Anyway, give us your opinion and we will give you a better response at the beginning of next week.

**Comment by klues** [comment]
I think the best would be a separate action like `fix grammar using ARASAAC API` which then can be added or not added in order to perform grammar correction before sharing/copying or not. It's most clear and flexible. Probably also an action under "Collect element action".

**Comment by arasaac-dga** [comment]
> I think the best would be a separate action like `fix grammar using ARASAAC API` which then can be added or not added in order to perform grammar correction before sharing/copying or not. It's most clear and flexible. Probably also an action under "Collect element action".

Yes, it makes sense but the problem is the delay in response. We have tested adding two actions as we have commented and due the delay in response from server when we have copied or shared the phrase was without fix. If you integrate the correction with the action, you can control the delay and wait to execute the action until you receive a response from the server.  Another problem is that actually "Speak collect element content" in collect elements actions fix the phrase before speak (which is the best and more transparent) and would be confusing for users to add a new subaction"fix grammar using ARASAAC API" as action. 

On the other hand in the Matrix Element (Send collect phrase (text + image) or (only text)) would be necessary that the correction work automatically as in the actual Collect Element Action (Speak collect element content) and only send the phrase to the chat when is fixed because is an image and cannot be modified after receive the correction server response.

**Comment by ms-mialingvo** [comment]
> I tried it on Edge / Windows - for me it worked 🤷

I just noticed I was having the same issue in Firefox iOS (working fine on Firefox Mac) when accessing gridbeta on the browser app directly but that it works when accessing it through the bookmark/"app-icon" on the desktop 🤷 Can't test Edge at the moment but maybe it will be the same there.

> background of shared image is white and width only as wide as the content - also better if content too long - still shows all elements in the screenshot

Works :)

> added two new actions for "Collect element action"

Not working on iPad (and Mac), browser-independently... On Firefox Mac I get "Invalid URI. Load of media resource  failed." in the developer console when clicking on the button.

**Comment by klues** [comment]
Updated [docs for Matrix messenger](https://github.com/asterics/AsTeRICS-Grid/blob/master/docs/documentation_user/13_matrix-messenger-integration.md).

> added two new actions for "Collect element action"
> Not working on iPad (and Mac), browser-independently...

For me at least sharing works on iPad Pro, iOs 18.3.2. Copying I couldn't test because I found nothing where I could paste the image on the iPad ;)

**Comment by klues** [comment]
> If you integrate the correction with the action, you can control the delay and wait to execute the action until you receive a response from the server.

I can also let the action "copy screenshot" or "share" wait until the other action ("get grammar correction from ARASAAC") is finished.

> Another problem is that actually "Speak collect element content" in collect elements actions fix the phrase before speak (which is the best and more transparent) and would be confusing for users to add a new subaction"fix grammar using ARASAAC API" as action.

In "Speak collect element content" I would leave it as-is, so that it automatically does the correction. I would just add another separate action for fixing grammar in order to be able to do it before copy image / share if it's intended.

> On the other hand in the Matrix Element (Send collect phrase (text + image) or (only text)) would be necessary that the correction work automatically as in the actual Collect Element Action

OK, I could add that.

**Comment by arasaac-dga** [comment]
> I can also let the action "copy screenshot" or "share" wait until the other action ("get grammar correction from ARASAAC") is finished.

No, you can launch it at the momment because is useful if, for example, somebody add it as separate button (and previosully press the PLAY button) as is and if somebody ask it we can comment it that is in development the new action.

> In "Speak collect element content" I would leave it as-is, so that it automatically does the correction

Yes, please. Don't modify this action. As is now is perfect. 

> I would just add another separate action for fixing grammar in order to be able to do it before copy image / share if it's intended.

In this case, the most important would be control the response from correction grammar server and that the system don't execute the next action in list (copy or share or waht ever other action) until the server has send the phrase fixed. Now, as we have commented we have tryed to combine 2 actions (Speak collect element content and Copy to Clipboard or Share) and doesn't work because the second action is executed before the server come back the phrase. So, if you develop the new action "get grammar correction from ARASAAC" some kind of control in actions flow is needed. 

> OK, I could add that.

Thanks so much. In this case is really important that works as actual "Speak collect element content".

**Comment by ms-mialingvo** [comment]
> For me at least sharing works on iPad Pro, iOs 18.3.2. Copying I couldn't test because I found nothing where I could paste the image on the iPad ;)

Am I doing something wrong then or missing something? 

![Image](https://github.com/user-attachments/assets/30894871-ae29-41cd-ad8d-30b668392062)

**Comment by klues** [comment]
> Thanks so much. In this case is really important that works as actual "Speak collect element content".

OK, for "send matrix message" actions (image or text) it's now the same as for "speak collect element content". If activated ARASAAC grammar correction is done before.

Regarding the other (copy / share) I'll leave it as-is for now and only implement some additional action if needed (but I doubt it).

> Am I doing something wrong then or missing something?

No, I don't think so. For me this action is showing the sharing dialog on iOS 18. But you must have something in the collect bar to be able to share something.
I think the "copy image to clipboard" does not work on iOS, I don't know why.

**Comment by klues** [comment]
I've added a spanish version of the pictogram communicator here: https://gridbeta.asterics-foundation.org/?gridset_filename=matrix_messenger_demo.grd.json

(use in private tab to get new version)

**Comment by ms-mialingvo** [comment]
> No, I don't think so. For me this action is showing the sharing dialog on iOS 18. But you must have something in the collect bar to be able to share something.

Weird. I have an iPad 9th generation with iOS 18.0.1. Made sure the browsers were updated just in case but that didn't make a difference. Copy text does work.

**Comment by arasaac-dga** [comment]
> OK, for "send matrix message" actions (image or text) it's now the same as for "speak collect element content". If activated ARASAAC grammar correction is done before.

Great!!! Thanks. We have tested and work fine.

> Regarding the other (copy / share) I'll leave it as-is for now and only implement some additional action if needed (but I doubt it).

Ok. We agree.

> I've added a spanish version of the pictogram communicator here (for testing ARASAAC API integration): https://gridbeta.asterics-foundation.org/?gridset_filename=matrix_messenger_demo.grd.json

Thanks. 

For us, you can launch it with new improvements.

**Comment by arasaac-dga** [comment]
> [@arasaac-dga](https://github.com/arasaac-dga) I assume you tested on Android? Does image share/copy work for you in any browsers other than Samsung Internet?

No, we have tested on Android, as usually, with Chrome and Edge (the most common browsers). In both everything work as expected. 

We have tested with Firefox and Samsumg Internet:

**+ Firefox:** none of both options work (share and copy image to clipboard) :-(
**+ Samsumg Internet:** sahre works, copy to clipboard no :-(

**Comment by ms-mialingvo** [comment]
> No, we have tested on Android, as usually, with Chrome and Edge (the most common browsers). In both everything work as expected.

I hadn't checked Chrome, it works there indeed. For me it's not working in Edge, neither sharing not copying...

**Comment by klues** [comment]
I also did some tests.

Windows:
* Chrome: both work
* Firefox: only copy works, share not supported: https://caniuse.com/web-share
* Edge: both work

Android:
* Chrome: both work
* Firefox: none works, strange because at least share should work in latest FF: https://caniuse.com/web-share

iOS:
* Safari: only share works (but in share you can select copy)
* Firefox: only share works (but in share you can select copy)

I'll leave it as-is for now, if something is important that should work, please add a new issue.

**Comment by klues** [comment]
released with https://github.com/asterics/AsTeRICS-Grid/releases/tag/release-2025-05-06-11.21%2F%2B0200
